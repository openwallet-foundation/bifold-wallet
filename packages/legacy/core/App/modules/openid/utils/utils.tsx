import { getDomainFromUrl } from '@credo-ts/core'
import { Attribute, Field } from '@hyperledger/aries-oca/build/legacy'
import { W3cCredentialDisplay } from '../types'

/**
 * Converts a camelCase string to a sentence format (first letter capitalized, rest in lower case).
 * i.e. sanitizeString("helloWorld")  // returns: 'Hello world'
 */
export function sanitizeString(str: string) {
  const result = str.replace(/([a-z0-9])([A-Z])/g, '$1 $2').replaceAll('_', ' ')
  let words = result.split(' ')
  words = words.map((word, index) => {
    if (index === 0 || word.toUpperCase() === word) {
      return word.charAt(0).toUpperCase() + word.slice(1)
    }
    return word.charAt(0).toLowerCase() + word.slice(1)
  })
  return words.join(' ')
}

export function getHostNameFromUrl(url: string) {
  try {
    return getDomainFromUrl(url)
  } catch (error) {
    throw new Error(`Error getting hostname from url: ${error}`)
  }
}

export const buildFieldsFromOpenIDTemplate = (data: { [key: string]: unknown }): Array<Field> => {
  const fields = []
  for (const key of Object.keys(data)) {
    // omit id and type
    if (key === 'id' || key === 'type') continue

    let pushedVal: string | number | null = null
    if (typeof data[key] === 'string' || typeof data[key] === 'number') {
      pushedVal = data[key] as string | number | null
    }
    fields.push(new Attribute({ name: key, value: pushedVal }))
  }
  return fields
}

export function formatDate(input: string | Date): string {
  const date = input instanceof Date ? input : new Date(input)
  return date.toLocaleString('en-US', {
    year: 'numeric',
    month: 'long',
    day: 'numeric',
    hour: '2-digit',
    minute: '2-digit',
    hour12: true,
  })
}

export const getAttributeField = (display: W3cCredentialDisplay, searchKey: string): Attribute | undefined => {
  for (const [key, value] of Object.entries(display.attributes)) {
    if (searchKey === key) {
      return new Attribute({
        name: key,
        value: value as string | number | null,
        mimeType: typeof value === 'number' ? 'text/number' : 'text/plain',
      })
    }
  }
  return undefined
}
