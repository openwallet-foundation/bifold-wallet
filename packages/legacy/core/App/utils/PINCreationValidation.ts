import { PINValidationRules } from '../types/security'

const consecutiveSeriesOfThree = new RegExp(/012|123|234|345|456|567|678|789|987|876|765|654|543|432|321|210/)
const evenNumberSeries = new RegExp('(13579)')
const oddNumberSeries = new RegExp('(02468)')
const isNumber = new RegExp('^[0-9]+$')
const crossNumberPattern = ['159753', '159357', '951357', '951753', '357159', '357951', '753159', '753951']

export enum PINError {
  CrossPatternValidation = 'CrossPatternValidation',
  OddOrEvenSequenceValidation = 'OddOrEvenSequenceValidation',
  NoRepetitionOfTheSameNumbersValidation = 'NoRepetitionOfTheSameNumbersValidation',
  MaxAdjacentNumberRepetitionValidation = 'MaxAdjacentNumberRepetitionValidation',
  NoRepetitionOfTheTwoSameNumbersValidation = 'NoRepetitionOfTheTwoSameNumbersValidation',
  NoSeriesOfNumbersValidation = 'NoSeriesOfNumbersValidation',
  PINOnlyContainDigitsValidation = 'PINOnlyContainDigitsValidation',
  PINTooShortValidation = 'PINTooShortValidation',
  PINTooLongValidation = 'PINTooLongValidation',
}

export interface PINValidationsType {
  isInvalid: boolean
  errorName: PINError
}

export const PINCreationValidations = (PIN: string, PINRules: PINValidationRules) => {
  const PINValidations: PINValidationsType[] = []
  
  if (PINRules.no_cross_pattern) {
    PINValidations.push({
      isInvalid: crossNumberPattern.includes(PIN),
      errorName: PINError.CrossPatternValidation,
    } as PINValidationsType)
  }
  if (PINRules.no_even_or_odd_series_of_numbers) {
    PINValidations.push({
      isInvalid: evenNumberSeries.test(PIN) || oddNumberSeries.test(PIN),
      errorName: PINError.OddOrEvenSequenceValidation,
    } as PINValidationsType)
  }

  if (1 == PINRules.no_repeated_numbers) {
    const repetitionPattern = new RegExp(/(\d)\1{1,}/)
    PINValidations.push({
      isInvalid: repetitionPattern.test(PIN),
      errorName: PINError.NoRepetitionOfTheSameNumbersValidation
    } as PINValidationsType)
  } else if (1 < PINRules.no_repeated_numbers){
    const repetitionPattern = new RegExp(String.raw`(\d)\1{${PINRules.no_repeated_numbers},}`, "g")
    PINValidations.push({
      isInvalid: repetitionPattern.test(PIN),
      errorName: PINError.MaxAdjacentNumberRepetitionValidation
    } as PINValidationsType)
  }

  if (PINRules.no_repetition_of_the_two_same_numbers) {
    let noRepetitionOfTheTwoSameNumbers = new RegExp(/([0-9][0-9])\1{1,}/)
    if (typeof PINRules.no_repetition_of_the_two_same_numbers === 'number') {
      noRepetitionOfTheTwoSameNumbers = new RegExp(
        `([0-9][0-9])\\1{${PINRules.no_repetition_of_the_two_same_numbers - 1},}`
      )
    }
    PINValidations.push({
      isInvalid: noRepetitionOfTheTwoSameNumbers.test(PIN),
      errorName: PINError.NoRepetitionOfTheTwoSameNumbersValidation,
    } as PINValidationsType)
  }
 
  if (PINRules.no_series_of_numbers) {
    PINValidations.push({
      isInvalid: consecutiveSeriesOfThree.test(PIN),
      errorName: PINError.NoSeriesOfNumbersValidation,
    } as PINValidationsType)
  }
  if (PINRules.only_numbers) {
    PINValidations.push({
      isInvalid: !isNumber.test(PIN),
      errorName: PINError.PINOnlyContainDigitsValidation,
    } as PINValidationsType)
  }

  PINValidations.push({
    isInvalid: PIN.length < PINRules.min_length || PIN.length > PINRules.max_length,
    errorName: PIN.length <= PINRules.max_length ? PINError.PINTooShortValidation : PINError.PINTooLongValidation,
  } as PINValidationsType)

  return PINValidations
}
