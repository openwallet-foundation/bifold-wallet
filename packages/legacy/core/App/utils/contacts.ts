import { BasicMessageRecord, ConnectionRecord, CredentialExchangeRecord, ProofExchangeRecord } from '@credo-ts/core'

import { BifoldAgent } from './agent'

interface ConnectionWithMessages {
  conn: ConnectionRecord
  msgs: (BasicMessageRecord | CredentialExchangeRecord | ProofExchangeRecord)[]
}

interface ConnectionWithLatestMessage {
  conn: ConnectionRecord
  latestMsg: BasicMessageRecord | CredentialExchangeRecord | ProofExchangeRecord
}

/**
 * Function to fetch contacts (connections) in order of latest chat message without using hooks
 * @param agent - Credo agent
 * @returns ConnectionRecord[] sorted by most recent message
 */
export const fetchContactsByLatestMessage = async (agent: BifoldAgent, connections: ConnectionRecord[]): Promise<ConnectionRecord[]> => {
  const connectionsWithMessages = await Promise.all<ConnectionWithMessages>(
    connections.map(
      async (conn: ConnectionRecord): Promise<ConnectionWithMessages> => ({
        conn,
        msgs: [
          ...(await agent.basicMessages.findAllByQuery({ connectionId: conn.id })),
          ...(await agent.proofs.findAllByQuery({ connectionId: conn.id })),
          ...(await agent.credentials.findAllByQuery({ connectionId: conn.id })),
        ],
      })
    )
  )

  const connectionsWithLatestMessage: ConnectionWithLatestMessage[] = connectionsWithMessages.map((pair) => {
    return {
      conn: pair.conn,
      latestMsg: pair.msgs.reduce(
        (acc, cur) => {
          const accDate = acc.updatedAt || acc.createdAt
          const curDate = cur.updatedAt || cur.createdAt
          return accDate > curDate ? acc : cur
        },
        // Initial value if no messages exist for this connection is a placeholder with the date the connection was created
        { createdAt: pair.conn.createdAt } as BasicMessageRecord | CredentialExchangeRecord | ProofExchangeRecord
      ),
    }
  })

  return connectionsWithLatestMessage
    .sort(
      (a, b) =>
        new Date(b.latestMsg.updatedAt || b.latestMsg.createdAt).valueOf() -
        new Date(a.latestMsg.updatedAt || a.latestMsg.createdAt).valueOf()
    )
    .map((pair) => pair.conn)
}
