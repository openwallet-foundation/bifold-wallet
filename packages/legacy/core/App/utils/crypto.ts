import argon2 from 'react-native-argon2'

export const hashPIN = async (PIN: string, salt: string): Promise<string> => {
  try {
    const result = await argon2(PIN, salt, {})
    const { rawHash } = result

    return rawHash
  } catch (error) {
    throw new Error(`Error generating hash for PIN ${String((error as Error)?.message ?? error)}`)
  }
}
