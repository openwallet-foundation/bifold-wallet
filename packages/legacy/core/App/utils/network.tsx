import TcpSocket from 'react-native-tcp-socket'

import pools from '../configs/ledgers/indy'
import { GenesisTransaction } from '../types/genesis'

export const canConnectToHost = async (host: { host: string; port: number }): Promise<boolean> =>
  new Promise((resolve) => {
    const socketTimeoutInMs = 3000
    const client = TcpSocket.createConnection(host, () => {
      resolve(true)
      client.destroy()
    })

    // Other events that can be safely be ignored. See the
    // library for more details:
    // https://www.npmjs.com/package/react-native-tcp-socket

    client.on('error', () => {
      client.destroy()
      resolve(false)
    })

    client.on('timeout', () => {
      client.destroy()
      client.removeAllListeners()

      resolve(false)
    })

    client.setTimeout(socketTimeoutInMs)
  })

export const fetchLedgerNodes = (indyNamespace = 'sovrin'): Array<{ host: string; port: number }> => {
  const [pool] = pools.filter((p) => p.indyNamespace === indyNamespace)
  if (!pool) {
    return []
  }

  const genesisTransactionsAsString = pool.genesisTransactions
  let genesisTransactions: Array<GenesisTransaction> = []

  if (genesisTransactionsAsString) {
    try {
      genesisTransactions = genesisTransactionsAsString.split('\n').map((g) => JSON.parse(g))
    } catch (error: unknown) {
      return []
    }
  }

  const nodes = genesisTransactions.map((g) => {
    return { host: g.txn.data.data.client_ip, port: parseInt(g.txn.data.data.client_port) }
  })

  return nodes
}
