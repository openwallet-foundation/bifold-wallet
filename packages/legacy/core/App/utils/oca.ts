import { CredentialExchangeRecord } from '@credo-ts/core'
import { AnonCredsProofRequestTemplatePayloadData, CredentialSharedProofData } from '@hyperledger/aries-bifold-verifier'
import {
  Attribute,
  CredentialOverlay,
  Field,
  OCABundleResolveAllParams,
  OCABundleResolverType,
  Predicate,
} from '@hyperledger/aries-oca/build/legacy'
import { W3cCredentialDisplay } from '../modules/openid/types'
import { BrandingOverlay } from '@hyperledger/aries-oca'

export const buildFieldsFromAnonCredsCredential = (credential: CredentialExchangeRecord): Array<Field> => {
  return credential?.credentialAttributes?.map((attr) => new Attribute(attr)) || []
}

export const buildFieldsFromW3cCredsCredential = (value: W3cCredentialDisplay): Array<Field> => {
  return (
    Object.entries(value.attributes).map(
      ([key, value]) =>
        new Attribute({
          name: key,
          value: value as string | number | null,
          mimeType: typeof value === 'number' ? 'text/number' : 'text/plain',
        })
    ) || []
  )
}

export const buildFieldsFromAnonCredsProofRequestTemplate = (
  data: AnonCredsProofRequestTemplatePayloadData
): Array<Field> => {
  const fields = []
  if (data.requestedAttributes) {
    for (const item of data.requestedAttributes) {
      if (item.name) {
        fields.push(new Attribute({ name: item.name, value: null, ...item }))
      }
      if (item.names) {
        for (const name of item.names) {
          fields.push(new Attribute({ name, value: null, ...item }))
        }
      }
    }
  }
  if (data.requestedPredicates) {
    for (const item of data.requestedPredicates) {
      fields.push(
        new Predicate({
          pType: item.predicateType,
          pValue: item.predicateValue,
          ...item,
        })
      )
    }
  }
  return fields
}

export const buildFieldsFromSharedAnonCredsProof = (data: CredentialSharedProofData): Array<Field> => {
  const fields = []
  for (const attribute of data.sharedAttributes) {
    fields.push(
      new Attribute({
        name: attribute.name,
        value: attribute.value,
      })
    )
  }
  for (const attributesGroup of data.sharedAttributeGroups) {
    for (const attribute of attributesGroup.attributes) {
      fields.push(
        new Attribute({
          name: attribute.name,
          value: attribute.value,
        })
      )
    }
  }
  for (const predicate of data.resolvedPredicates) {
    fields.push(
      new Predicate({
        name: predicate.name,
        pType: predicate.predicateType,
        pValue: predicate.predicateValue,
        satisfied: true,
      })
    )
  }
  return fields
}

export const buildOverlayFromW3cCredential = async ({
  credentialDisplay,
  language,
  resolver,
}: {
  credentialDisplay: W3cCredentialDisplay
  language: string
  resolver: OCABundleResolverType
}): Promise<CredentialOverlay<BrandingOverlay>> => {
  const params: OCABundleResolveAllParams = {
    identifiers: {
      schemaId: '',
      credentialDefinitionId: credentialDisplay.id,
    },
    meta: {
      alias: credentialDisplay.display.issuer.name,
      credConnectionId: undefined,
      credName: credentialDisplay.display.name,
    },
    attributes: buildFieldsFromW3cCredsCredential(credentialDisplay),
    language,
  }

  const bundle = await resolver.resolveAllBundles(params)

  return {
    ...(bundle as CredentialOverlay<BrandingOverlay>),
    presentationFields: buildFieldsFromW3cCredsCredential(credentialDisplay),
  }
}
