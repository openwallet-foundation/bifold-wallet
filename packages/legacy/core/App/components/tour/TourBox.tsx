import React, { ReactElement, ReactNode, useCallback, useEffect, useState } from 'react'
import { useTranslation } from 'react-i18next'
import { StyleSheet, Text, TouchableOpacity, View } from 'react-native'
import Icon from 'react-native-vector-icons/MaterialIcons'

import { hitSlop } from '../../constants'
import { useTheme } from '../../contexts/theme'
import { RenderProps } from '../../contexts/tour/tour-context'
import { testIdWithKey } from '../../utils/testable'

enum PaginationDotTypes {
  Filled,
  Unfilled,
}

export interface TourBoxProps extends RenderProps {
  children?: ReactNode
  /**
   * Optionally hide the left button
   */
  hideLeft?: boolean
  /**
   * Optionally hide the right button
   */
  hideRight?: boolean
  /**
   * Text for the left button
   */
  leftText?: string
  /**
   * Text for the right button
   */
  rightText?: string
  /**
   * Callback for when the left button is pressed.
   */
  onLeft?: () => void
  /**
   * Callback for when the right button is pressed.
   */
  onRight?: () => void
  title?: string
  /**
   * The 1-indexed step it is on
   */
  stepOn?: number
  /**
   * The 1-indexed number of steps in the current tour
   */
  stepsOutOf?: number
}

/**
 * A built-in TourBox component which can be used as a tooltip container for
 * each step. While it's somewhat customizable, it's not required and can be
 * replaced by your own component.
 *
 * @param props the component props
 * @returns A TourBox React element
 */
export function TourBox(props: TourBoxProps): ReactElement {
  const { t } = useTranslation()
  const {
    leftText = t('Tour.Back'),
    rightText = t('Tour.Next'),
    title,
    hideLeft,
    hideRight,
    onLeft,
    onRight,
    children,
    stop,
    stepOn,
    stepsOutOf,
  } = props
  const { TextTheme, ColorPallet, OnboardingTheme } = useTheme()
  const iconSize = 30
  const dismissIconName = 'clear'
  const iconColor = ColorPallet.notification.infoIcon
  const [paginationDots, setPaginationDots] = useState<PaginationDotTypes[]>([])
  const [xPos, setXPos] = useState(0)
  const swipeThreshold = 75

  const styles = StyleSheet.create({
    container: {
      backgroundColor: ColorPallet.notification.info,
      borderColor: ColorPallet.notification.infoBorder,
      borderRadius: 5,
      borderWidth: 1,
      padding: 20,
      flex: 1,
      margin: 'auto',
    },
    headerContainer: {
      flexDirection: 'row',
      justifyContent: 'space-between',
    },
    headerTextContainer: {
      flexGrow: 1,
    },
    headerText: {
      ...TextTheme.headingThree,
      alignSelf: 'flex-start',
      color: ColorPallet.notification.infoText,
    },
    dismissIcon: {
      alignSelf: 'flex-end',
    },
    body: {
      flex: 1,
      marginVertical: 16,
    },
    footerContainer: {
      flexDirection: 'row',
      justifyContent: 'space-around',
    },
    navText: {
      ...TextTheme.bold,
      color: ColorPallet.brand.primary,
    },
    pagerContainer: {
      flexDirection: 'row',
      alignSelf: 'center',
    },
    pagerDot: {
      ...OnboardingTheme.pagerDot,
      borderWidth: 1,
      borderStyle: 'solid',
      width: 10,
      height: 10,
      borderRadius: 5,
      marginHorizontal: 5,
    },
    pagerDotActive: {
      backgroundColor: OnboardingTheme.pagerDotActive.color,
    },
    pagerDotInactive: {
      backgroundColor: 'transparent',
    },
  })

  const handleLeft = useCallback((): void => {
    onLeft?.()
  }, [onLeft])

  const handleRight = useCallback((): void => {
    onRight?.()
  }, [onRight])

  useEffect(() => {
    const arr: PaginationDotTypes[] = []
    if (typeof stepOn === 'number' && typeof stepsOutOf === 'number') {
      for (let i = 1; i <= stepsOutOf; i++) {
        if (i === stepOn) {
          arr.push(PaginationDotTypes.Filled)
        } else {
          arr.push(PaginationDotTypes.Unfilled)
        }
      }
    }

    setPaginationDots(arr)
  }, [stepOn, stepsOutOf])

  return (
    <View
      style={styles.container}
      onTouchStart={(e) => setXPos(e.nativeEvent.pageX)}
      onTouchEnd={(e) => {
        // if swipe right
        if (xPos - e.nativeEvent.pageX > swipeThreshold) {
          handleRight()

          // if swipe left
        } else if (e.nativeEvent.pageX - xPos > swipeThreshold) {
          handleLeft()
        }
      }}
    >
      <View style={styles.headerContainer}>
        <View style={styles.headerTextContainer}>
          <Text
            allowFontScaling={false}
            style={styles.headerText}
            testID={testIdWithKey('HeaderText')}
            accessibilityRole="header"
          >
            {title}
          </Text>
        </View>
        <View style={styles.dismissIcon}>
          <TouchableOpacity
            onPress={stop}
            testID={testIdWithKey('Close')}
            accessibilityLabel={t('Global.Close')}
            accessibilityRole={'button'}
            hitSlop={hitSlop}
          >
            <Icon name={dismissIconName} size={iconSize} color={iconColor} />
          </TouchableOpacity>
        </View>
      </View>
      <View style={styles.body}>{children}</View>

      {(!hideLeft || !hideRight) && (
        <View style={styles.footerContainer}>
          <View>
            {!hideLeft && (
              <TouchableOpacity
                accessibilityLabel={leftText}
                accessibilityRole={'button'}
                testID={testIdWithKey('Back')}
                onPress={handleLeft}
                hitSlop={hitSlop}
              >
                <Text allowFontScaling={false} style={styles.navText}>
                  {leftText}
                </Text>
              </TouchableOpacity>
            )}
          </View>
          <View style={styles.pagerContainer}>
            {paginationDots.map((dot, index) =>
              dot === PaginationDotTypes.Filled ? (
                <View key={index} style={[styles.pagerDot, styles.pagerDotActive]}></View>
              ) : (
                <View key={index} style={[styles.pagerDot, styles.pagerDotInactive]}></View>
              )
            )}
          </View>
          <View>
            {!hideRight && (
              <TouchableOpacity
                accessibilityLabel={rightText}
                accessibilityRole={'button'}
                testID={testIdWithKey('Next')}
                onPress={handleRight}
                hitSlop={hitSlop}
              >
                <Text allowFontScaling={false} style={styles.navText}>
                  {rightText}
                </Text>
              </TouchableOpacity>
            )}
          </View>
        </View>
      )}
    </View>
  )
}
