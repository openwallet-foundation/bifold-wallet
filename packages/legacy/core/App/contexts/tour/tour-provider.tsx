import React, { forwardRef, Ref, useCallback, useImperativeHandle, useMemo, useState } from 'react'
import { ColorValue, LayoutRectangle } from 'react-native'

import { TourOverlay } from '../../components/tour/TourOverlay'
import { BaseTourID, ChildFn, TourID } from '../../types/tour'
import { isChildFunction } from '../../utils/helpers'

import {
  BackdropPressBehavior,
  OSConfig,
  Tour,
  TourContext,
  TourCtx,
  TourStep,
  ORIGIN_SPOT,
  Tours,
} from './tour-context'

export interface TourProviderProps {
  children: React.ReactNode | ChildFn<Tour>
  /**
   * Define if the animations in the tour should use the native driver or not.
   * A boolean can be used to apply the same value to both Android and iOS, or
   * an object with `android` and `ios` keys can be used to define a value for
   * each OS.
   *
   * @default false
   */
  nativeDriver?: boolean | OSConfig<boolean>
  /**
   * Sets the default behavior of pressing the tour's backdrop. You can use
   * either one of the following values:
   * - A callback function with the {@link Tour} options object in the
   * first argument. This allows more franular control over the tour.
   * - The `continue` literal string, which is a shortcut to move to the next
   * step, and stop the tour on the last step.
   * - the `stop` literal string, which is a shortcut to stop the tour.
   *
   * **NOTE:** You can also override this behavior on each step configuration.
   */
  onBackdropPress?: BackdropPressBehavior
  /**
   * The color of the overlay of the tour.
   *
   * @default black
   */
  overlayColor?: ColorValue
  /**
   * The opacity applied to the overlay of the tour (between 0 to 1).
   *
   * @default 0.45
   */
  overlayOpacity?: number
  /**
   * The list of steps for the home tour.
   */
  tours: Tours
}

const TourProviderComponent = (props: TourProviderProps, ref: Ref<Tour>) => {
  const {
    children,
    onBackdropPress,
    overlayColor = 'black',
    overlayOpacity = 0.45,
    tours,
    nativeDriver = false,
  } = props

  const [currentTour, setCurrentTour] = useState<TourID>(BaseTourID.HomeTour)
  const [currentStep, setCurrentStep] = useState<number>()
  const [spot, setSpot] = useState(ORIGIN_SPOT)

  const renderStep = useCallback(
    (index: number): void | Promise<void> => {
      if (tours[currentTour]?.[index] !== undefined) {
        setCurrentStep(index)
      }
    },
    [currentTour, tours]
  )

  const changeSpot = useCallback((newSpot: LayoutRectangle): void => {
    setSpot(newSpot)
  }, [])

  const start = useCallback(
    (tourId: TourID): void => {
      setCurrentTour(tourId)
      renderStep(0)
    },
    [renderStep]
  )

  const stop = useCallback((): void => {
    setCurrentStep(undefined)
    setSpot(ORIGIN_SPOT)
  }, [])

  const next = useCallback((): void => {
    if (currentTour && currentStep !== undefined && tours[currentTour]) {
      currentStep === tours[currentTour].length - 1 ? stop() : renderStep(currentStep + 1)
    }
  }, [stop, renderStep, currentStep, currentTour, tours])

  const previous = useCallback((): void => {
    if (currentStep !== undefined && currentStep > 0) {
      renderStep(currentStep - 1)
    }
  }, [renderStep, currentStep])

  const tourStep = useMemo((): TourStep => {
    return tours?.[currentTour]?.[currentStep ?? 0] ?? { Render: () => <></> }
  }, [currentTour, currentStep, tours])

  const tour = useMemo(
    (): TourCtx => ({
      changeSpot,
      currentTour,
      currentStep,
      next,
      previous,
      spot,
      start,
      stop,
      tours,
    }),
    [changeSpot, currentTour, currentStep, next, previous, spot, start, stop, tours]
  )

  useImperativeHandle(ref, () => ({
    currentTour,
    currentStep,
    changeSpot,
    next,
    previous,
    start,
    stop,
  }))

  return (
    <TourContext.Provider value={tour}>
      {isChildFunction(children) ? <TourContext.Consumer>{children}</TourContext.Consumer> : <>{children}</>}

      <TourOverlay
        color={overlayColor}
        currentStep={currentStep}
        currentTour={currentTour}
        changeSpot={changeSpot}
        backdropOpacity={overlayOpacity}
        onBackdropPress={onBackdropPress}
        spot={spot}
        tourStep={tourStep}
        nativeDriver={nativeDriver}
      />
    </TourContext.Provider>
  )
}

export const TourProvider = forwardRef<Tour, TourProviderProps>(TourProviderComponent)
