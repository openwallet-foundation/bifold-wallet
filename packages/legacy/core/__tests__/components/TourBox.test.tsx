import { fireEvent, render } from '@testing-library/react-native'
import React from 'react'

import { TourBox } from '../../App/components/tour/TourBox'
import { TourProvider } from '../../App/contexts/tour/tour-provider'
import { testIdWithKey } from '../../App/utils/testable'
import { toursMock } from '../fixtures/tours'

describe('TourBox Component', () => {
  test('Renders properly with defaults', () => {
    const previous = jest.fn()
    const next = jest.fn()
    const stop = jest.fn()
    const tree = render(
      <TourProvider tours={toursMock} overlayColor={'gray'} overlayOpacity={0.7}>
        <TourBox
          title={'Title'}
          leftText={'Left'}
          rightText={'Right'}
          onLeft={previous}
          onRight={next}
          currentTour={'homeTourSteps'}
          currentStep={0}
          previous={previous}
          stop={stop}
          next={next}
        />
      </TourProvider>
    )

    const { getByTestId } = tree
    const leftButton = getByTestId(testIdWithKey('Back'))
    const rightButton = getByTestId(testIdWithKey('Next'))
    fireEvent(leftButton, 'press')
    expect(previous).toHaveBeenCalledTimes(1)
    fireEvent(rightButton, 'press')
    expect(next).toHaveBeenCalledTimes(1)

    expect(tree).toMatchSnapshot()
  })
})
