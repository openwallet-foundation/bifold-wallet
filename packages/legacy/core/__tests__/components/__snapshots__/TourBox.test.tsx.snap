// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TourBox Component Renders properly with defaults 1`] = `
<View
  onTouchEnd={[Function]}
  onTouchStart={[Function]}
  style={
    Object {
      "backgroundColor": "#313132",
      "borderColor": "#0099FF",
      "borderRadius": 5,
      "borderWidth": 1,
      "flex": 1,
      "margin": "auto",
      "padding": 20,
    }
  }
>
  <View
    style={
      Object {
        "flexDirection": "row",
        "justifyContent": "space-between",
      }
    }
  >
    <View
      style={
        Object {
          "flexGrow": 1,
        }
      }
    >
      <Text
        accessibilityRole="header"
        allowFontScaling={false}
        style={
          Object {
            "alignSelf": "flex-start",
            "color": "#FFFFFF",
            "fontSize": 26,
            "fontWeight": "bold",
          }
        }
        testID="com.ariesbifold:id/HeaderText"
      >
        Title
      </Text>
    </View>
    <View
      style={
        Object {
          "alignSelf": "flex-end",
        }
      }
    >
      <View
        accessibilityLabel="Global.Close"
        accessibilityRole="button"
        accessibilityState={
          Object {
            "busy": undefined,
            "checked": undefined,
            "disabled": undefined,
            "expanded": undefined,
            "selected": undefined,
          }
        }
        accessibilityValue={
          Object {
            "max": undefined,
            "min": undefined,
            "now": undefined,
            "text": undefined,
          }
        }
        accessible={true}
        collapsable={false}
        focusable={true}
        hitSlop={
          Object {
            "bottom": 44,
            "left": 44,
            "right": 44,
            "top": 44,
          }
        }
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Object {
            "opacity": 1,
          }
        }
        testID="com.ariesbifold:id/Close"
      >
        <Text
          allowFontScaling={false}
          selectable={false}
          style={
            Array [
              Object {
                "color": "#0099FF",
                "fontSize": 30,
              },
              undefined,
              Object {
                "fontFamily": "Material Icons",
                "fontStyle": "normal",
                "fontWeight": "normal",
              },
              Object {},
            ]
          }
        >
          î…Œ
        </Text>
      </View>
    </View>
  </View>
  <View
    style={
      Object {
        "flex": 1,
        "marginVertical": 16,
      }
    }
  />
  <View
    style={
      Object {
        "flexDirection": "row",
        "justifyContent": "space-around",
      }
    }
  >
    <View>
      <View
        accessibilityLabel="Left"
        accessibilityRole="button"
        accessibilityState={
          Object {
            "busy": undefined,
            "checked": undefined,
            "disabled": undefined,
            "expanded": undefined,
            "selected": undefined,
          }
        }
        accessibilityValue={
          Object {
            "max": undefined,
            "min": undefined,
            "now": undefined,
            "text": undefined,
          }
        }
        accessible={true}
        collapsable={false}
        focusable={true}
        hitSlop={
          Object {
            "bottom": 44,
            "left": 44,
            "right": 44,
            "top": 44,
          }
        }
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Object {
            "opacity": 1,
          }
        }
        testID="com.ariesbifold:id/Back"
      >
        <Text
          allowFontScaling={false}
          style={
            Object {
              "color": "#42803E",
              "fontSize": 18,
              "fontWeight": "bold",
            }
          }
        >
          Left
        </Text>
      </View>
    </View>
    <View
      style={
        Object {
          "alignSelf": "center",
          "flexDirection": "row",
        }
      }
    />
    <View>
      <View
        accessibilityLabel="Right"
        accessibilityRole="button"
        accessibilityState={
          Object {
            "busy": undefined,
            "checked": undefined,
            "disabled": undefined,
            "expanded": undefined,
            "selected": undefined,
          }
        }
        accessibilityValue={
          Object {
            "max": undefined,
            "min": undefined,
            "now": undefined,
            "text": undefined,
          }
        }
        accessible={true}
        collapsable={false}
        focusable={true}
        hitSlop={
          Object {
            "bottom": 44,
            "left": 44,
            "right": 44,
            "top": 44,
          }
        }
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Object {
            "opacity": 1,
          }
        }
        testID="com.ariesbifold:id/Next"
      >
        <Text
          allowFontScaling={false}
          style={
            Object {
              "color": "#42803E",
              "fontSize": 18,
              "fontWeight": "bold",
            }
          }
        >
          Right
        </Text>
      </View>
    </View>
  </View>
</View>
`;
